File: server/src/volume.js
'use strict';

const { db } = require('./db');

/**
 * §4D Transaction Volume by Market. Stacked by Market, split Low Coverage vs
 * Land, bucketed by month of `record_date` (the underlying market-activity
 * date, not upload date — see README for why: grouping by upload batch would
 * conflate "when I happened to import a file" with "when the deal actually
 * happened"). Same shape for availabilities (new listings appearing).
 */
function monthBucket(iso) {
  return iso ? iso.slice(0, 7) : 'unknown';
}

function buildVolume(recordType) {
  const rows = db.prepare('SELECT * FROM listings WHERE record_type = ?').all(recordType);
  const buckets = new Map(); // month -> market -> type -> {count, sqft, acres}

  for (const r of rows) {
    const month = monthBucket(r.record_date);
    if (!buckets.has(month)) buckets.set(month, {});
    const b = buckets.get(month);
    const key = `${r.market}__${r.type_of_land_normalized || 'Unclassified'}`;
    if (!b[key]) b[key] = { market: r.market, type: r.type_of_land_normalized || 'Unclassified', count: 0, sqft: 0, acres: 0 };
    b[key].count += 1;
    if (typeof r.sqft === 'number') b[key].sqft += r.sqft;
    if (typeof r.acres === 'number') b[key].acres += r.acres;
  }

  const months = [...buckets.keys()].sort();
  const series = months.map((month) => ({ month, ...buckets.get(month) }));
  return { recordType, months, series };
}

function getVolume() {
  return {
    transactions: buildVolume('transaction'),
    availabilities: buildVolume('availability'),
  };
}

module.exports = { getVolume };

File: streamlit_app/lib/volume.py
"""Transaction Volume by Market — ported from volume.js. Bucketed by month
of record_date (the underlying market-activity date, not upload date).
"""

from . import db


def _month_bucket(iso):
    return iso[:7] if iso else "unknown"


def _build_volume(record_type):
    rows = [dict(r) for r in db.query("SELECT * FROM listings WHERE record_type = ?", (record_type,))]
    buckets = {}
    for r in rows:
        month = _month_bucket(r["record_date"])
        buckets.setdefault(month, {})
        key = f"{r['market']}__{r['type_of_land_normalized'] or 'Unclassified'}"
        b = buckets[month].setdefault(key, {"market": r["market"], "type": r["type_of_land_normalized"] or "Unclassified", "count": 0, "sqft": 0, "acres": 0})
        b["count"] += 1
        if isinstance(r["sqft"], (int, float)):
            b["sqft"] += r["sqft"]
        if isinstance(r["acres"], (int, float)):
            b["acres"] += r["acres"]

    months = sorted(buckets.keys())
    series = [{"month": m, **buckets[m]} for m in months]
    return {"record_type": record_type, "months": months, "series": series}


def get_volume():
    return {"transactions": _build_volume("transaction"), "availabilities": _build_volume("availability")}
